File: api/src/org/ufomatics/archipelago/model/MetaFunctionImpl.java
package org.ufomatics.archipelago.model;

import org.ufomatics.archipelago.model.MetaType;
import org.ufomatics.archipelago.model.MetaField;
import org.ufomatics.archipelago.model.MetaFunction;

import java.io.Serializable;
import java.util.Set;
import java.util.HashSet;

/**
 * @author <a href="mailto:rr0@rr0.org">Jerome Beau</a>
 * @version $revision$
 */
public class MetaFunctionImpl implements MetaFunction, Serializable {
    private String name;
    private Set<MetaField> parameters = new HashSet<MetaField>();
    private MetaType returnType;

    public MetaFunctionImpl() {
    }

    public void setName(String name) {
        this.name = name;
    }

    public void addParameter(MetaField metaField) {
        parameters.add (metaField);
    }

    public MetaField createParameter() {
        return new MetaFieldImpl();
    }

    public String getName() {
        return name;
    }

    public void add(MetaField field) {
        parameters.add (field);
    }

    public void setParameters(Set<MetaField> parameters) {
        this.parameters = parameters;
    }

    public Set<MetaField> getParameters() {
        return parameters;
    }

    public MetaType getReturnType() {
        return returnType;
    }

    public void setReturnType(MetaType returnType) {
        this.returnType = returnType;
    }

    public String toString() {
        return getName() + ": " + getReturnType();
    }
}

File: src/archipelago/api/model/MetaFunctionImpl.ts
import {MetaFunction} from "archipelago/api/model/MetaFunction"
import {MetaField} from "archipelago/api/model/MetaField"
import {MetaType} from "archipelago/api/model/MetaType"
import {MetaFieldImpl} from "archipelago/api/model/MetaFieldImpl"
import {VOID} from "archipelago/api/model/MetaTypeImpl"
import {JSerializable} from "ts-jsdk"

/**
 *
 */
export class MetaFunctionImpl implements MetaFunction, JSerializable {
  private name = ""
  private parameters = new Set<MetaField>()
  private returnType: MetaType = VOID

  constructor() {
  }

  setName(name: string): void {
    this.name = name
  }

  addParameter(metaField: MetaField): void {
    this.parameters.add(metaField)
  }

  createParameter(name: string, type: MetaType): MetaField {
    return new MetaFieldImpl(name, type)
  }

  getName(): string {
    return this.name
  }

  add(field: MetaField): void {
    this.parameters.add(field)
  }

  setParameters(parameters: Set<MetaField>): void {
    this.parameters = parameters
  }

  getParameters(): Set<MetaField> {
    return this.parameters
  }

  getReturnType(): MetaType {
    return this.returnType
  }

  setReturnType(returnType: MetaType): void {
    this.returnType = returnType
  }

  toString(): string {
    return this.getName() + ": " + this.getReturnType()
  }
}
